test/SecTester.Scan.Tests/ScanTests.cs

```diff
@@ -346,4 +327,115 @@ public async Task DisposeAsync_DeleteScanThrows_DoesNothing()
     await act.Should().NotThrowAsync();
     await _scans.Received(1).DeleteScan(ScanId);
   }
+
+  [Theory]
+  [MemberData(nameof(DoneStatuses))]
+  public async Task Expect_GivenPredicateFinalStateReached_Returns(ScanStatus scanStatus)
+  {
+    // arrange
+    _scans.GetScan(ScanId).Returns(
+        new ScanState(ScanStatus.Running), new ScanState(scanStatus));
+    _predicate(Arg.Is<Scan>(x => x.Id == ScanId)).Returns(false);
+
+    // act
+    var act = () => _sut.Expect(_predicate);
+
+    // assert
+    await act.Should().NotThrowAsync();
+    await _predicate.Received(2)(Arg.Any<Scan>());
+  }
+
+  [Fact]
+  public async Task Expect_GivenPredicateCancelledByTimeout_Returns()
+  {
+    // arrange
+    var sut = new Scan(ScanId, _scans, _logger,
+      new ScanOptions()
+      {
+        PollingInterval = TimeSpan.FromMilliseconds(50),
+        Timeout = TimeSpan.Zero
+      });
+    await using var _ = sut;
+
+    _scans.GetScan(ScanId).Returns(new ScanState(ScanStatus.Running));
+    _predicate(Arg.Is<Scan>(x => x.Id == ScanId)).Returns(false);
+
+    // act
+    var act = () => sut.Expect(_predicate);
+
+    // assert
+    await act.Should().NotThrowAsync();
+    await _predicate.Received(1)(Arg.Any<Scan>());
+  }
+
+  [Theory]
+  [MemberData(nameof(ActiveStatuses))]
+  public async Task Expect_GivenPredicateConditionSatisfied_Returns(ScanStatus scanStatus)
+  {
+    // arrange
+    _scans.GetScan(ScanId).Returns(new ScanState(scanStatus));
+    _predicate(Arg.Is<Scan>(x => x.Id == ScanId)).Returns(true);
+
+    // act
+    var act = () => _sut.Expect(_predicate);
+
+    // assert
+    await act.Should().NotThrowAsync();
+    await _predicate.Received(1)(Arg.Any<Scan>());
+  }
+
+  [Theory]
+  [MemberData(nameof(ActiveStatuses))]
+  public async Task Expect_GivenPredicateWithStatusMatching_Returns(ScanStatus scanStatus)
+  {
+    // arrange
+    var satisfyingScanState = new ScanState(scanStatus)
+    {
+      IssuesBySeverity = new[] { new IssueGroup(1, Severity.High) }
+    };
+
+    async Task<bool> Predicate(Scan scan)
```

Comment:
So, why don't add 3rd level of predicate?) You should use NSubstitute to operate, verify, and capture arguments